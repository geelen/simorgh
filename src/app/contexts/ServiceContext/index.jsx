import React from 'react';
import { node, string } from 'prop-types';
import services from '#lib/config/services/loadableConfig';
import { getVariant } from '../../lib/utilities/variantHandler';
import variantPropType from '../../models/propTypes/variants';

export const ServiceContext = React.createContext({});

export const ServiceContextProvider = ({ children, service, variant }) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={getVariant({ service, variant })}
    >
      {children}
    </LoadableContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
  variant: variantPropType,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
  variant: 'default',
};
