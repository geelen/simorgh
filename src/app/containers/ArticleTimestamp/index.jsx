import React, { useContext } from 'react';
import { number, bool, string } from 'prop-types';
import Timestamp from '@bbc/psammead-timestamp-container';
import {
  PopOutGridItemMedium,
  GridItemConstrainedMedium,
} from '#lib/styledGrid';
import { ServiceContext } from '#contexts/ServiceContext';
import { formatDateNumeric } from './timeFormats';
import {
  isFirstRelative,
  isLastRelative,
  formatType,
  isValidDateTime,
} from './helpers';

const ArticleTimestamp = ({
  firstPublished,
  lastPublished,
  popOut,
  minutesTolerance,
  className,
}) => {
  const {
    articleTimestampPrefix,
    datetimeLocale,
    script,
    service,
    timezone,
    altCalendar,
  } = useContext(ServiceContext);

  if (!isValidDateTime(firstPublished) || !isValidDateTime(lastPublished)) {
    return null;
  }

  const timestampProps = {
    dateTimeFormat: formatDateNumeric,
    script,
    locale: datetimeLocale,
    service,
    timezone,
    altCalendar,
  };
  const firstPublishedProps = {
    timestamp: firstPublished,
    format: formatType({ firstPublished }),
    isRelative: isFirstRelative(firstPublished, lastPublished),
  };

  const lastPublishedProps = {
    timestamp: lastPublished,
    format: formatType({ lastPublished, firstPublished }),
    isRelative: isLastRelative(lastPublished),
    prefix: articleTimestampPrefix,
  };

  const Wrapper = popOut ? PopOutGridItemMedium : GridItemConstrainedMedium;

  const timeDifferenceMinutes = (lastPublished - firstPublished) / 1000 / 60;

  return (
    <Wrapper className={className}>
      <Timestamp {...timestampProps} {...firstPublishedProps} />
      {timeDifferenceMinutes > minutesTolerance && (
        <Timestamp {...timestampProps} {...lastPublishedProps} />
      )}
    </Wrapper>
  );
};

ArticleTimestamp.propTypes = {
  firstPublished: number.isRequired,
  lastPublished: number.isRequired,
  popOut: bool,
  minutesTolerance: number,
  // eslint-disable-next-line react/require-default-props
  className: string,
};

ArticleTimestamp.defaultProps = {
  popOut: true,
  minutesTolerance: 0,
};

export default ArticleTimestamp;
