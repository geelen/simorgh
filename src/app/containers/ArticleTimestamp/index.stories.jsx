import React, { useEffect } from 'react';
import { storiesOf } from '@storybook/react';
import { element } from 'prop-types';
import timemachine from 'timemachine';
import ArticleTimestamp from '.';
import { timestampGenerator } from './testHelpers';
import { ServiceContextProvider } from '#contexts/ServiceContext';
import services from '#testHelpers/serviceConfigs';

timemachine.reset();

const threeHoursAgo = timestampGenerator({ hours: 3 });
const fiveHoursAgo = timestampGenerator({ hours: 5 });
const elevenHoursAgo = timestampGenerator({ hours: 11 });
const twelveHoursAgo = timestampGenerator({ hours: 12 });
const twentyFourHoursAgo = timestampGenerator({
  hours: 24,
  seconds: 1,
});
const twoDaysAgo = timestampGenerator({ days: 2 });
const threeDaysAgo = timestampGenerator({ days: 3 });

// eslint-disable-next-line react/prop-types
const WrappedArticleTimestamp = ({ service, ...rest }) => (
  <ServiceContextProvider service={service || 'news'}>
    <ArticleTimestamp {...rest} />
  </ServiceContextProvider>
);

const stories = storiesOf(
  'Containers|Article/Article Timestamp',
  module,
).addParameters({ chromatic: { disable: true } });

stories.addDecorator(story => {
  // This affects the global Date object for the storybook application, to ensure consistency in chromaticQA testing.
  const WithTimemachine = ({ children }) => {
    useEffect(() => {
      timemachine.config({
        dateString: 'Friday, 9 August 2019 14:04:14',
        timestamp: 1565359454,
      });
      return () => {
        timemachine.reset();
      };
    });
    return <>{children}</>;
  };
  WithTimemachine.propTypes = {
    children: element.isRequired,
  };
  return <WithTimemachine>{story()}</WithTimemachine>;
});

stories.add('default', () => (
  <WrappedArticleTimestamp
    firstPublished={1530947227000}
    lastPublished={1552666749637}
  />
));

Object.keys(services)
  .filter(service => service !== 'default')
  .forEach(service => {
    stories.add(service, () => (
      <WrappedArticleTimestamp
        firstPublished={1530947227000}
        lastPublished={1552666749637}
        service={service}
      />
    ));
  });

stories
  .add(
    'lastPublished === firstPublished and firstPublished < 10 hours ago',
    () => (
      <WrappedArticleTimestamp
        firstPublished={threeHoursAgo}
        lastPublished={threeHoursAgo}
      />
    ),
  )
  .add(
    'lastPublished === firstPublished and firstPublished today and > 10 hours ago',
    () => (
      <WrappedArticleTimestamp
        firstPublished={elevenHoursAgo}
        lastPublished={elevenHoursAgo}
      />
    ),
  )
  .add(
    'lastPublished === firstPublished and firstPublished before today',
    () => (
      <WrappedArticleTimestamp
        firstPublished={twentyFourHoursAgo}
        lastPublished={twentyFourHoursAgo}
      />
    ),
  )
  .add(
    'lastPublished today < 10 hours ago and firstPublished today < 10 hours ago',
    () => (
      <WrappedArticleTimestamp
        firstPublished={fiveHoursAgo}
        lastPublished={threeHoursAgo}
      />
    ),
  )
  .add(
    'lastPublished today more than 10 hours ago and firstPublished today more than 10 hours ago',
    () => (
      <WrappedArticleTimestamp
        firstPublished={twelveHoursAgo}
        lastPublished={elevenHoursAgo}
      />
    ),
  )
  .add(
    'lastPublished before today and firstPublished same day as lastPublished',
    () => (
      <WrappedArticleTimestamp
        firstPublished={threeDaysAgo}
        lastPublished={threeDaysAgo}
      />
    ),
  )
  .add(
    'lastPublished before today, !==firstPublished day and firstPublished before today',
    () => (
      <WrappedArticleTimestamp
        firstPublished={threeDaysAgo}
        lastPublished={twoDaysAgo}
      />
    ),
  )
  .add(
    'lastPublished today and more than 10 hours ago and firstPublished before today',
    () => (
      <WrappedArticleTimestamp
        firstPublished={threeDaysAgo}
        lastPublished={elevenHoursAgo}
      />
    ),
  );
