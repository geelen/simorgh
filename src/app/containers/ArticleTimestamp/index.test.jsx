import React from 'react';
import { render, mount } from 'enzyme';
import {
  isNull,
  suppressPropWarnings,
  shouldMatchSnapshot,
} from '@bbc/psammead-test-helpers';
import ArticleTimestamp from '.';
import {
  timestampGenerator,
  sameDayTimestampsGenerator,
  isBritishSummerTime,
} from './testHelpers';
import { ServiceContextProvider } from '#contexts/ServiceContext';

const regexDate = /[0-9]{1,2} \w+ [0-9]{4}/;
const regexDatetime = /[0-9]{1,2} \w+ [0-9]{4}[,] [0-9]{2}[:][0-9]{2} \w+/;

const regexUpdatedDatetime = /Updated [0-9]{1,2} \w+ [0-9]{4}[,] [0-9]{2}[:][0-9]{2} \w+/;
const regexUpdatedDate = /^Updated [0-9]{1,2} \w+ [0-9]{4}$/;

const firstChild = wrapper => wrapper[0].children[0].data;
const secondChild = wrapper => wrapper[1].children[0].data;

const renderedTimestamps = jsx => render(jsx).get(0).children; // helper as output is wrapped in a grid

// eslint-disable-next-line react/prop-types
const WrappedArticleTimestamp = ({ service, ...props }) => (
  <ServiceContextProvider service={service}>
    <ArticleTimestamp {...props} />
  </ServiceContextProvider>
);

WrappedArticleTimestamp.defaultProps = {
  service: 'news',
};

describe('ArticleTimestamp', () => {
  let originalDate;
  const inBritishSummerTime = isBritishSummerTime(Date.now());
  beforeEach(() => {
    originalDate = Date.now;
  });

  afterEach(() => {
    Date.now = originalDate;
  });

  shouldMatchSnapshot(
    "should render a 'created' Timestamp correctly",
    <WrappedArticleTimestamp
      firstPublished={1530947227000}
      lastPublished={1530947227000}
    />,
  );

  shouldMatchSnapshot(
    "should render both a 'created' and an 'updated' Timestamp correctly",
    <WrappedArticleTimestamp
      firstPublished={1530947227000}
      lastPublished={1552666749637}
    />,
  );

  describe('daylight savings time', () => {
    const daylightSavingsBehaviour = ({ descriptor, date, longName }) => {
      it(`should produce ${descriptor} as a descriptor when in ${longName}`, () => {
        const [
          moreThanTenHoursAgo,
          mockCurrentTimestamp,
        ] = sameDayTimestampsGenerator({
          date,
          intervals: [{ hours: 10, seconds: 25 }],
        });
        Date.now = jest.fn(() => mockCurrentTimestamp);
        const renderedWrapper = renderedTimestamps(
          <WrappedArticleTimestamp
            firstPublished={moreThanTenHoursAgo}
            lastPublished={moreThanTenHoursAgo}
          />,
        );

        expect(renderedWrapper.length).toEqual(1);
        expect(renderedWrapper[0].children[0].data).toMatch(regexDatetime);
        expect(renderedWrapper[0].children[0].data).toContain(descriptor);
      });
    };

    const testValues = [
      {
        descriptor: 'BST',
        date: '2017-05-31',
        longName: 'British Summer Time',
      },
      {
        descriptor: 'GMT',
        date: '2017-01-01',
        longName: 'Greenwich Mean Time',
      },
    ];

    for (let i = 0; i < testValues.length; i += 1) {
      daylightSavingsBehaviour(testValues[i]);
    }
  });

  describe('with no data', () => {
    suppressPropWarnings(['firstPublished', 'undefined']);
    suppressPropWarnings(['lastPublished', 'undefined']);
    isNull('should return null', <WrappedArticleTimestamp />);
  });

  describe('with invalid data', () => {
    isNull(
      'should return null',
      <WrappedArticleTimestamp
        firstPublished={8640000000000001}
        lastPublished={8640000000000001}
      />,
    );
  });

  it('should render one timestamp with relative time when firstPublished < 10 hours ago and lastPublished === firstPublished', () => {
    const [threeHoursAgo, mockCurrentTimestamp] = sameDayTimestampsGenerator({
      intervals: [{ hours: 3 }],
    });
    Date.now = jest.fn(() => mockCurrentTimestamp);
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={threeHoursAgo}
        lastPublished={threeHoursAgo}
      />,
    );
    expect(renderedWrapper.length).toEqual(1);
    expect(firstChild(renderedWrapper)).toEqual('3 hours ago');
  });

  it('should render one timestamp with date & time when firstPublished today and > 10 hours ago and lastPublished === firstPublished', () => {
    const [elevenHoursAgo, mockCurrentTimestamp] = sameDayTimestampsGenerator({
      intervals: [{ hours: 11 }],
    });
    Date.now = jest.fn(() => mockCurrentTimestamp);
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={elevenHoursAgo}
        lastPublished={elevenHoursAgo}
      />,
    );
    expect(renderedWrapper.length).toEqual(1);
    expect(firstChild(renderedWrapper)).toMatch(regexDatetime);
  });

  it('should render one timestamp with date when firstPublished before today and lastPublished === firstPublished', () => {
    const twentyFourHoursAgo = timestampGenerator({
      hours: 24,
      seconds: 1,
    });
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={twentyFourHoursAgo}
        lastPublished={twentyFourHoursAgo}
      />,
    );
    expect(renderedWrapper.length).toEqual(1);
    expect(firstChild(renderedWrapper)).toMatch(regexDate);
  });

  it('should render one timestamp with date when firstPublished before today and lastPublished was published less than 1 minute after firstPublished', () => {
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={1530947280000}
        lastPublished={1530947286000}
        minutesTolerance={1}
      />,
    );
    expect(renderedWrapper.length).toEqual(1);
    expect(firstChild(renderedWrapper)).toMatch(regexDate);
  });

  it('should render two timestamps with date when firstPublished before today and lastPublished was published more than 1 minute after firstPublished', () => {
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={1530947280000}
        lastPublished={1530947406000}
        minutesTolerance={1}
      />,
    );
    expect(renderedWrapper.length).toEqual(2);
    expect(firstChild(renderedWrapper)).toMatch(regexDate);
  });

  it('should render two timestamps - published: date & time, updated: relative when both are today and < 10 hours ago', () => {
    const [
      fiveHoursAgo,
      threeHoursAgo,
      mockCurrentTimestamp,
    ] = sameDayTimestampsGenerator({
      intervals: [{ hours: 5 }, { hours: 3 }],
    });
    Date.now = jest.fn(() => mockCurrentTimestamp);
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={fiveHoursAgo}
        lastPublished={threeHoursAgo}
      />,
    );
    expect(renderedWrapper.length).toEqual(2);
    expect(firstChild(renderedWrapper)).toMatch(regexDatetime);
    expect(secondChild(renderedWrapper)).toMatch('3 hours ago');
  });

  it('should render two timestamps - published: date & time, updated: date & time when both are today and > 10 hours ago', () => {
    const [
      twelveHoursAgo,
      elevenHoursAgo,
      mockCurrentTimestamp,
    ] = sameDayTimestampsGenerator({
      intervals: [{ hours: 12 }, { hours: 11 }],
    });
    Date.now = jest.fn(() => mockCurrentTimestamp);
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={twelveHoursAgo}
        lastPublished={elevenHoursAgo}
      />,
    );
    expect(renderedWrapper.length).toEqual(2);
    expect(firstChild(renderedWrapper)).toMatch(regexDatetime);
    expect(secondChild(renderedWrapper)).toMatch(regexUpdatedDatetime);
  });

  it('should render two timestamps - published: date, updated: date when firstPublished before today and lastPublished before today, but not same day as firstPublished', () => {
    const threeDaysAgo = timestampGenerator({
      days: 3,
    });
    const twoDaysAgo = timestampGenerator({ days: 2 });
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={threeDaysAgo}
        lastPublished={twoDaysAgo}
      />,
    );
    expect(renderedWrapper.length).toEqual(2);
    expect(firstChild(renderedWrapper)).toMatch(regexDate);
    expect(secondChild(renderedWrapper)).toMatch(regexUpdatedDate);
  });

  it('should render two timestamps - published: date, updated: date when firstPublished before today and lastPublished today and > 10 hrs ago', () => {
    const timestamp = inBritishSummerTime ? 1496235600000 : 1483275600000;
    Date.now = jest.fn(() => timestamp);
    const threeDaysAgo = timestampGenerator({
      days: 3,
    });
    const elevenHoursAgo = timestampGenerator({ hours: 11 });
    const renderedWrapper = renderedTimestamps(
      <WrappedArticleTimestamp
        firstPublished={threeDaysAgo}
        lastPublished={elevenHoursAgo}
      />,
    );
    expect(renderedWrapper.length).toEqual(2);
    expect(firstChild(renderedWrapper)).toMatch(regexDate);
    expect(secondChild(renderedWrapper)).toMatch(regexUpdatedDate);
  });

  describe('With different timezones', () => {
    it('should show the correct local date', () => {
      const props = {
        firstPublished: 1565380800000,
        lastPublished: 1565380800000,
      };
      const newsContainer = mount(
        <WrappedArticleTimestamp {...props} service="news" />,
      );

      const time = newsContainer.find('time').props().dateTime;
      expect(time).toEqual('2019-08-09');

      const bengaliContainer = mount(
        <WrappedArticleTimestamp {...props} service="bengali" />,
      );
      const bengaliTime = bengaliContainer.find('time').props().dateTime;
      expect(bengaliTime).toEqual('2019-08-10');
    });
  });
});
