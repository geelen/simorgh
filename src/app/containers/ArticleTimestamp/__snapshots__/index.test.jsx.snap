// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleTimestamp should render a 'created' Timestamp correctly 1`] = `
.c1 {
  font-size: 0.875rem;
  line-height: 1.125rem;
  color: #6E6E73;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  padding-bottom: 0.25rem;
}

.c1:last-child {
  padding-bottom: 1rem;
}

.c0 {
  grid-column: 1 / span 6;
}

@media (min-width:20rem) and (max-width:37.4375rem) {
  .c1 {
    font-size: 0.875rem;
    line-height: 1.125rem;
  }
}

@media (min-width:37.5rem) {
  .c1 {
    font-size: 0.8125rem;
    line-height: 1rem;
  }
}

@media (max-width:25rem) {
  .c0 {
    padding: 0 0.5rem;
  }
}

@media (min-width:25rem) and (max-width:62.9375rem) {
  .c0 {
    padding: 0 1rem;
  }
}

@media (min-width:37.5rem) {
  .c0 {
    grid-column: 1 / span 5;
    max-width: 83.33%;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c0 {
    grid-column: 3 / span 5;
    max-width: 37.75rem;
  }
}

@media (min-width:80rem) {
  .c0 {
    grid-column: 2 / span 4;
    max-width: 38.5rem;
  }
}

@supports (display:grid) {
  .c0 {
    max-width: initial;
  }
}

<div
  class="c0"
>
  <time
    class="c1"
    datetime="2018-07-07"
  >
    7 July 2018
  </time>
</div>
`;

exports[`ArticleTimestamp should render both a 'created' and an 'updated' Timestamp correctly 1`] = `
.c1 {
  font-size: 0.875rem;
  line-height: 1.125rem;
  color: #6E6E73;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  padding-bottom: 0.25rem;
}

.c1:last-child {
  padding-bottom: 1rem;
}

.c0 {
  grid-column: 1 / span 6;
}

@media (min-width:20rem) and (max-width:37.4375rem) {
  .c1 {
    font-size: 0.875rem;
    line-height: 1.125rem;
  }
}

@media (min-width:37.5rem) {
  .c1 {
    font-size: 0.8125rem;
    line-height: 1rem;
  }
}

@media (max-width:25rem) {
  .c0 {
    padding: 0 0.5rem;
  }
}

@media (min-width:25rem) and (max-width:62.9375rem) {
  .c0 {
    padding: 0 1rem;
  }
}

@media (min-width:37.5rem) {
  .c0 {
    grid-column: 1 / span 5;
    max-width: 83.33%;
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c0 {
    grid-column: 3 / span 5;
    max-width: 37.75rem;
  }
}

@media (min-width:80rem) {
  .c0 {
    grid-column: 2 / span 4;
    max-width: 38.5rem;
  }
}

@supports (display:grid) {
  .c0 {
    max-width: initial;
  }
}

<div
  class="c0"
>
  <time
    class="c1"
    datetime="2018-07-07"
  >
    7 July 2018
  </time>
  <time
    class="c1"
    datetime="2019-03-15"
  >
    Updated 
    15 March 2019
  </time>
</div>
`;
